sequencer/src/state.rs

```diff
@@ -1,47 +1,141 @@
-use crate::{chain_variables::ChainVariables, Error, Header, Payload};
+use crate::{
+    block::{
+        BlockMerkleCommitment, BlockMerkleTree, FeeAccount, FeeAmount, FeeMerkleCommitment,
+        FeeMerkleTree,
+    },
+    Error, Header, Payload,
+};
 use commit::{Commitment, Committable};
-use ethers::prelude::BlockNumber;
 use hotshot::traits::State as HotShotState;
-use hotshot_types::data::ViewNumber;
+use hotshot_types::data::{BlockError, ViewNumber};
+use jf_primitives::merkle_tree::{
+    AppendableMerkleTreeScheme, MerkleTreeScheme, UniversalMerkleTreeScheme,
+};
 use serde::{Deserialize, Serialize};
-use std::fmt::Debug;
 
-#[derive(Default, Serialize, Deserialize, Clone, Debug, Hash, PartialEq, Eq)]
-pub struct State {
-    chain_variables: ChainVariables,
-    ethereum_block_tag: BlockNumber,
+#[derive(Hash, Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct ValidatedState {
+    /// Frontier of Block Merkle Tree
+    pub block_merkle_tree: BlockMerkleTree,
+    /// Fee Merkle Tree
+    pub fee_merkle_tree: FeeMerkleTree,
 }
 
-impl HotShotState for State {
-    type Error = Error;
+impl Default for ValidatedState {
+    fn default() -> Self {
+        let block_merkle_tree =
+            BlockMerkleTree::from_elems(32, Vec::<Commitment<Header>>::new()).unwrap();
+        let fee_merkle_tree =
+            FeeMerkleTree::from_kv_set(20, Vec::<(FeeAccount, FeeAmount)>::new()).unwrap();
+        Self {
+            block_merkle_tree,
+            fee_merkle_tree,
+        }
+    }
+}
+
+impl ValidatedState {
+    fn validate_proposal(&self, parent: &Header, proposal: &Header) -> anyhow::Result<Self> {
```

Comment:
This seemed like a good place to do the validation. All of this could just as easily be in `self.validate_and_apply()` but I liked keeping the trait implementation clean for the moment. Contents here are the previous `_validate_proposal` standalone function plus addition validation against fee and block merkle trees in state. The signature differs from that of `_validate_proposal` due to addition content. Maybe I am deviating from some original intention that I am not aware of? Original `_validate_proposal` returned `Result<BlockMerkleTree>`.

Reply:
Yeah I don't think there was any great intent in the original signature. I think the new signature is a result of just getting a clearer understanding of the problem. So I think this is the right direction 👍